File: app/src/main/java/org/inventivetalent/notificationlogger/database/Notification.java
package org.inventivetalent.notificationlogger.database;

import androidx.room.ColumnInfo;
import androidx.room.Entity;
import androidx.room.PrimaryKey;

import org.json.JSONObject;

import java.util.Date;

@Entity(tableName = "notifications")
public class Notification {


    @PrimaryKey(autoGenerate = true)
    public int id;


    @ColumnInfo
    public String action;

    @ColumnInfo
    public Date time;

    @ColumnInfo
    public int notificationId;

    @ColumnInfo
    public String key;

    @ColumnInfo
    public String tag;

    @ColumnInfo
    public String packageName;

    @ColumnInfo
    public String tickerText;

    @ColumnInfo(name = "when")
    public Date whenTime;

    @ColumnInfo
    public int number;

    @ColumnInfo
    public int visibility;

    @ColumnInfo
    public int priority;

    @ColumnInfo
    public String sound;

    @ColumnInfo
    public String vibrate;// Long array pattern stored as comma separated string

    @ColumnInfo
    public String channelId;

    @ColumnInfo
    public String channelName;

    @ColumnInfo
    public String channelDescription;

    @ColumnInfo
    public JSONObject extrasJson;

}

File: app/src/main/java/org/inventivetalent/notificationlogger/database/Notification.kt
package org.inventivetalent.notificationlogger.database

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.PrimaryKey
import org.json.JSONObject
import java.util.*

@Entity(tableName = "notifications")
class Notification {


    @PrimaryKey(autoGenerate = true)
    var id: Int = 0


    @ColumnInfo
    var action: String? = null

    @ColumnInfo
    var time: Date? = null

    @ColumnInfo
    var notificationId: Int = 0

    @ColumnInfo
    var key: String? = null

    @ColumnInfo
    var tag: String? = null

    @ColumnInfo
    var packageName: String? = null

    @ColumnInfo
    var tickerText: String? = null

    @ColumnInfo(name = "when")
    var whenTime: Date? = null

    @ColumnInfo
    var number: Int = 0

    @ColumnInfo
    var visibility: Int = 0

    @ColumnInfo
    var priority: Int = 0

    @ColumnInfo
    var interruptionFilter: Int = 0

    @ColumnInfo
    var removeReason: Int = 0

    @ColumnInfo
    var sound: String? = null

    @ColumnInfo
    var vibrate: String? = null// Long array pattern stored as comma separated string

    @ColumnInfo
    var category: String? = null

    @ColumnInfo
    var color: Int = 0

    @ColumnInfo
    var flags: Int = 0

    @ColumnInfo
    var channelId: String? = null

    @ColumnInfo
    var channelName: String? = null

    @ColumnInfo
    var channelDescription: String? = null

    @ColumnInfo
    var extrasJson: JSONObject? = null

    fun hasExtra(key:String):Boolean{
        return extrasJson != null && (extrasJson as JSONObject).has(key)
    }

    fun getExtraString(key:String, def:String?=null):String?{
        if (hasExtra(key)) {
            return extrasJson?.getString(key)
        }
        return def
    }

    fun getExtraInt(key:String, def:Int=0):Int?{
        if (hasExtra(key)) {
            return extrasJson?.getInt(key)
        }
        return def
    }

}
